### PR TITLE
stricter hidden type wf-check [based on #115008]

### LINKED ISSUE #114572: Incorrect lifetime bound check in async + impl_trait_in_assoc_type
I tried this code:

```rust
#![feature(type_alias_impl_trait)]
#![feature(impl_trait_in_assoc_type)]

use std::{fmt::Display, future::Future};

pub trait SharedDisplay: Display + Send + 'static {}

pub struct Message<T: AsRef<dyn SharedDisplay>> {
    inner: T,
}

pub trait AsyncPrinter: Send + Sync {
    type AsyncOutput<'a>: Future<Output = String> + Send
    where
        Self: 'a;

    fn to_string(&self, msg: Message<Box<dyn SharedDisplay>>) -> Self::AsyncOutput<'_>;
}

struct FooPrinter;

impl AsyncPrinter for FooPrinter {
    type AsyncOutput<'a> = impl Future<Output = String> + Send + 'a where Self: 'a;

    fn to_string(&self, msg: Message<Box<dyn SharedDisplay>>) -> Self::AsyncOutput<'_> {
        async move { msg.inner.to_string() }
    }
}
```

I expected to see this happen: *The code correctly compiles*

Instead, this happened: *The code fails to compile with lifetime bound check errors.*

### More information that might be helpful

I did a bisect and found that the above code compiles in `nightly-2023-03-14` but fails in `nightly-2023-03-15`, possibly due to https://github.com/rust-lang/rust/pull/108909 but I didn't dig deep to find the root cause.

Also, by changing the implementation of the `to_string` function of `FooPrinter` by wrapping the async block with a smart pointer (see below), the code successfully compiles

```rust
fn to_string(&self, msg: Message<Box<dyn SharedDisplay>>) -> Self::AsyncOutput<'_> {
    Box::pin(async move { msg.inner.to_string() })
}
```

Let me know if more info is needed, or actually my codes fails to compile by design.

### Meta

`rustc --version --verbose`:
```
rustc 1.73.0-nightly (474709a9a 2023-08-03)
binary: rustc
commit-hash: 474709a9a2a74a8bcf0055fadb335d0ca0d2d939
commit-date: 2023-08-03
host: x86_64-apple-darwin
release: 1.73.0-nightly
LLVM version: 16.0.5
```

<details><summary>Backtrace</summary>
<p>

```
error[E0478]: lifetime bound not satisfied
  --> src/lib.rs:23:28
   |
23 |     type AsyncOutput<'a> = impl Future<Output = String> + Send + 'a where Self: 'a;
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: lifetime parameter instantiated with the lifetime `'a` as defined here
  --> src/lib.rs:23:22
   |
23 |     type AsyncOutput<'a> = impl Future<Output = String> + Send + 'a where Self: 'a;
   |                      ^^
   = note: but lifetime parameter must outlive the static lifetime

```

</p>
</details>

### LINKED ISSUE #114728: RPIT hidden types can be ill-formed
The following UB shouldn't compile:
```rust
type Static<'a> = &'static &'a ();
trait Extend<'a> {
    fn extend(self, _: &'a str) -> &'static str;
}
impl<'a> Extend<'a> for Static<'a> {
    fn extend(self, s: &'a str) -> &'static str {
        s
    }
}
fn boom<'a>(arg: Static<'_>) -> impl Extend<'a> {
    arg
}
fn main() {
    let y = boom(&&()).extend(&String::from("temporary"));
    println!("{}", y);
}
```
This regressed in v1.61. I guess it is #94081.

Discovered when investigating #114572.

### PR DESCRIPTION
**Pull request by lcnr**
Original work by @aliemjay in #115008. A huge thanks to them for originally figuring out this approach :heart:

Fixes https://github.com/rust-lang/rust/issues/114728
Fixes https://github.com/rust-lang/rust/issues/114572

Instead of adding the `WellFormed` obligations when relating opaque types, we now always emit such an obligation when defining the hidden type.

This causes nested opaque types which aren't wf to error, see the comment below for the described impact. I believe this change to be desirable as it significantly reduces complexity by removing special-cases.

It also caused an issue with RPITIT: in defaulted trait methods, we add a `Projection(synthetic_assoc, rpit_of_trait_method)` clause to the `param_env`. This clause is not added to the `ParamEnv` of the nested coroutines. This caused a normalization failure in `fn check_coroutine_obligations` with the new solver. I fixed that by using the env of the typeck root instead.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/check/check.rs [hunk 1]
```diff
@@ -402,39 +402,29 @@ fn check_opaque_meets_bounds<'tcx>(
...
-            let _ = infcx.take_opaque_types();
-            return Ok(());
-        }
-        // Nested opaque types occur only in associated types:
-        // ` type Opaque<T> = impl Trait<&'static T, AssocTy = impl Nested>; `
-        // They can only be referenced as `<Opaque<T> as Trait<&'static T>>::AssocTy`.
-        // We don't have to check them here because their well-formedness follows from the WF of
-        // the projection input types in the defining- and use-sites.
-        hir::OpaqueTyOrigin::TyAlias { .. }
-            if tcx.def_kind(tcx.parent(def_id.to_def_id())) == DefKind::OpaqueTy => {}
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/check/check.rs:422]: I removed this, cc https://github.com/rust-lang/rust/pull/106038#discussion_r1133074120 @aliemjay

I do not think this special-case is worth it and it's fairly fragile, as we cannot check that nested opaque types are well formed while in the defining scope. While I am still positive that the previous behavior was sound, having to allow types which aren't well-formed adds complexity which I don't think is worth it here.

File: tests/ui/type-alias-impl-trait/wf-check-definition-site.rs
Comment (oli-obk) [inline tests/ui/type-alias-impl-trait/wf-check-definition-site.rs:1]: the comments in this test could benefit from some updating on what happens now (only `Opaque<'a> = Static<&'static str>` is the hidden type)

**Review by oli-obk — APPROVED** [review]
broke my brain for 20 mins, but makes sense now

**Review by compiler-errors — APPROVED** [review]

**Comment by aliemjay** [comment]
*Context*: we're discussing whether the following code is valid despite the fact that the hidden type `impl Sized` is not technically well-formed because `T: 'static` is not known in `test` fn. However, if we consider the full context where the nested opaque type is defined `impl Trait<&'static T, Assoc = impl Sized>`, I believe we can safely assume that `T: 'static` holds.
```rust
trait Id<X> { type Ty; }
impl<X> Id<X> for () { type Ty = X; }
fn test() -> impl Id<&'static T, Ty = impl Sized> {}
```

This was accepted as a valid defining use in #106038 and will be rejected with with this PR with the concern of increased complexity.

> This causes nested opaque types which aren't wf to error, see the comment below for the described impact. I believe this change to be desirable as it significantly reduces complexity by removing special-cases.

This breaking change is not necessary for the soundness fix here and thus I prefer to at least have it in a separate PR.

Regarding the complexity concern, I believe we can keep supporting this behavior in a less complicated way than in #115008. The previous implementation in #115008 was unnecessarily complex because I wanted to have a perf baseline by emitting the least possible number of WF obligations. Moreover, I think that nested opaque types are special enough to have their own `hir::OpaqueTyOrigin` variant and this would arguably make the added complexity more acceptable.

**Comment by aliemjay** [comment]
> This causes nested opaque types which aren't wf to error, see the comment below for the described impact. I believe this change to be desirable as it significantly reduces complexity by removing special-cases.

Oh, this would also disallow a more natural pattern that is much more likely to appear in the wild:
```rust
// This is rejected currently because of #96146 but even when this is fixed
// it would still be rejected by this PR with an error that requires `T: 'static` bound in `test`
fn id<T>(s: &T) -> &T { s }
fn test<T>() -> impl Fn(&T) -> impl Sized {
  id
}
```

**Comment by lcnr** [comment]
I have strong feelings here and believe we should not support this behavior.

---

Regarding the complexity. My main concern is not the behavior of `check_opaque_meets_bounds`. I strongly believe that we the following statement should be true:

**We should always be able to assume, and verify, that all types are well-formed in the current context.**

Allowing these patterns breaks this assumption at the point where we define the nested opaque type. This is fragile in two ways:

We will have to be careful when emitting WF constraints as shown in this PR to avoid this breakage going forward. It is incredibly subtle why my approach broke the `wf-nested.rs` tests while your approach in #115008 did not. Your approach only emits `WellFormed` obligations when relating opaques. We therefore avoid checking well-formedness of nested opaques as we eagerly replaced the nested opaque with an inference variable in `project_and_unify_type`: https://github.com/rust-lang/rust/blob/71a7b66f20c551f640f2f382bc7e7923ba0a5dab/compiler/rustc_trait_selection/src/traits/project.rs#L259-L269

This is not done in the new solver. We'd have to work around it there to continue to support this, for example by also eagerly replacing opaques with inference variables.

Even more importantly, we have to worry about accidentally leaking the non-wf type to somewhere which accepts well-formed types. We already have multiple unsoundnesses caused by a lack of WF checks, e.g. #98117. I do not want to make it more difficult to fix such issues as that may end up forcing nested opaques to be well-formed.

---

Supporting this pattern by amending the environment while defining nested opaque types seems a lot more acceptable. However, I believe that doing so is far from trivial.

---

Finally, I don't think this behavior is (very) desirable or consistent with the rest of Rust. I feel like "this was accepted as a valid defining use in #106038" is mostly irrelevant as the concern in https://github.com/rust-lang/rust/pull/106038#discussion_r1069351480 ended up mostly slipping through and there hasn't been any official team approval of this behavior.This diverges from the behavior of where-clauses where such reasoning is not supported:
```rust
trait Id<X> { type Ty; }
impl<X> Id<X> for () { type Ty = X; }

fn test<T>() 
where
    (): Id<&'static T>,
    // (): Id<&'static T, Ty = &'static T> {}
    // both of these bounds error
```

I do agree that supporting the following is generally desirable:
```rust
fn id<T>(s: &T) -> &T { s }
fn test<T>() -> impl Fn(&T) -> impl Sized {
  id
}
```
However, even this is imo not a strong argument to not land this PR as is. We currently [do not support defining higher-ranked opaque types **at all**](https://github.com/rust-lang/rust/pull/116935#issuecomment-1969536775). Trying to do so always errors. I will push against any such attempts until the new solver is exclusively used as I am very much unsure how to soundly handle this when using *semantic lookup* for the opaque type storage.

Once implied bounds are explicitly handled and checked (necessary to fix #25860), we can support the above pattern without needing to support non-wf types as the type would now only be used inside of the `Binder` with the right bounds.

---

> This breaking change is not necessary for the soundness fix here and thus I prefer to at least have it in a separate PR.

I don't. I consider the breakage for RPIT to have already been FCP'd as part of #115008.

> However this is still a regression:
> ```rust
> trait Id<X> { type Ty; }
> impl<X> Id<X> for () { type Ty = X; }
> fn test<T>() -> impl Id<&'static T, Ty = impl Sized> {}
> ```
> We should probably ignore the wf-check nested RPIT, similar to nested TAIT. 

We could separate the removal of the nested TAIT special-case into a separate PR, but given that the added WF-checks during borrowck reduce the breakage caused by that I don't think such a split is desirable. I personally don't have the energy and motivation to first land the approach of #115008 and then move the WF-checks, breaking `wf-nested.rs` in a separate PR.

**Comment by oli-obk** [comment]
The two regressions have similar up and down fluctuations over the last few weeks.

@rustbot label: +perf-regression-triaged

**Comment by oli-obk** [comment]
@bors r+

This may be stricter than necessary, but we can relax it later if we realize this is supportable in the new solver

**Comment by lcnr** [comment]
This checks that RPIT are well-formed in more cases, fixing a soundness issue.

**Pull request merged**